server/internal/infprocessor/infprocessor.go

```diff
@@ -33,18 +34,32 @@ type Task struct {
 
 // WaitForCompletion waits for the completion of the task.
 func (t *Task) WaitForCompletion(ctx context.Context) (*http.Response, error) {
-	log.Printf("Waiting for the completion of the task: %s\n", t.ID)
+	log.Printf("Waiting for the completion of the task (ID: %q)\n", t.ID)
 	select {
 	case <-ctx.Done():
+		// When a task result comes, the processor still attempts to
+		// write a response/error to a channel. We need to read
+		// from the channel to avoid a goroutine leak.
+		go t.discardResp()
 		return nil, ctx.Err()
 	case resp := <-t.RespCh:
 		return resp, nil
 	case err := <-t.ErrCh:
-		log.Printf("Task failed: %s\n", err)
+		log.Printf("Task (ID: %q) failed: %s\n", t.ID, err)
 		return nil, err
 	}
 }
 
+func (t *Task) discardResp() {
+	select {
```

Comment:
Question: does either respCh or errCh always return a value after a context is canceled?

Reply:
Yes, that's correct! 

- If the engine is able to correctly process a request, the server receives a response and write to `respCh`.
- if some error happens (e.g., engine is disconnected), an error is written to `errCh`.

For example, if an engine is disconnected, we call `RemoveEngine`, which writes an error to all in-progress tasks.

https://github.com/llm-operator/inference-manager/blob/ebe48f7c31014a3e83c972f4049b602150c8aff1/server/internal/infprocessor/infprocessor.go#L192